src/main/java/com/fasterxml/jackson/databind/jsontype/impl/TypeNameIdResolver.java

```diff
@@ -80,7 +82,16 @@ public static TypeNameIdResolver construct(MapperConfig<?> config, JavaType base
     @Override
     public String idFromValue(Object value)
     {
-        Class<?> cls = _typeFactory.constructType(value.getClass()).getRawClass();
+       return idFromClass(value.getClass());
+    }
+
+
+    protected String idFromClass(Class<?> clazz)
+    {
+        if(clazz==null){
```

Comment:
I understand the need, from perspective that while traversal is based on serialization, in many aspects it really relates more to deserialization because there is no access to instances just types.
I am just trying to see how pieces fit together: looking at code I can see how it relates to `JsonSerializer`, which is the entrypoint. I didn't think visitors exposed information directly.